### PR TITLE
Add normalized os and app attributes

### PR DESCRIPTION
**Pull request by jklukas**
Closes https://github.com/mozilla/gcp-ingestion/issues/478

### CODE REVIEW

File: ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java [hunk 1]
```diff
@@ -4,28 +4,54 @@
...
-   * <p>Reimplementation of https://github.com/mozilla-services/lua_sandbox_extensions/blob/sprint2018-6/moz_telemetry/modules/moz_telemetry/normalize.lua#L167-L178
+   * <p>Reimplementation of https://github.com/mozilla-services/lua_sandbox_extensions/blob/607b3c7b9def600ddbd373a14f67a55e1c69f7a1/moz_telemetry/modules/moz_telemetry/normalize.lua#L167-L178
    */
   public static String channel(String name) {
+    name = Strings.nullToEmpty(name);
     if (CHANNELS.contains(name)) {
       // We could lowercase before doing this check, but don't to maintain historical practice.
       return name;
-    } else if (name == null) {
-      return OTHER;
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java]: Changes to the `channel` method here are to better match the style used in the new methods introduced in this PR.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java [hunk 2]
```diff
@@ -39,4 +65,64 @@ public static String channel(String name) {
       return OTHER;
     }
   }
+
+  /**
+   * Normalize an operating system string to one of the three major desktop platforms,
+   * one of the two major mobile platforms, or "Other".
+   *
+   * <p>Reimplements and combines the logic of {@code os} and {@code mobile_os} from
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java]: We combine the desktop and mobile here to support the [propose common metadata structure](https://docs.google.com/document/d/1AsshL_QS1qRVXcHW5lz0jZeM-UKIWFl1oV3XJz7Xewo/edit?disco=AAAACsG0pZA&usp_dm=false&ts=5cb4b194). All tables loaded by the ingestion pipeline will have a `normalized_os` field.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java [hunk 3]
```diff
@@ -39,4 +65,64 @@ public static String channel(String name) {
...
+      return "";
+    }
+  }
+
+  /**
+   * Normalize app_name as reported by the various products, or "Other" for unrecognized prefixes.
+   *
+   * <p>Reimplements {@code mobile_app_name} from
+   * https://github.com/mozilla-services/lua_sandbox_extensions/blob/607b3c7b9def600ddbd373a14f67a55e1c69f7a1/moz_telemetry/modules/moz_telemetry/normalize.lua#L218-L230
+   * and adds a case for "Firefox" desktop app.
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java]: Likewise here, all tables loaded by ingestion will have a `normalized_app_name` field, so we support the desktop case along with mobile.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java [hunk 4]
```diff
@@ -4,28 +4,54 @@
...
 
+  private static final String WINDOWS = "Windows";
+  private static final String MAC = "Mac";
+  private static final String LINUX = "Linux";
+  private static final String IOS = "iOS";
+  private static final String ANDROID = "Android";
+
+  private static final List<String> APP_NAMES = ImmutableList.of(
+      // Make sure more-specific versions come before Zerda.
+      "Zerda_cn", "Zerda",
```
Comment (jklukas) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java]: I think there's a bug in [the existing lua implementation of app name normalization](https://github.com/mozilla-services/lua_sandbox_extensions/blob/607b3c7b9def600ddbd373a14f67a55e1c69f7a1/moz_telemetry/modules/moz_telemetry/normalize.lua#L218-L230) where Zerda is checked before Zerda_cn, so "Zerda_cn" will be normalized to "Zerda". Same for "Focus-TV" normalized to "Focus".

File: ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java [hunk 5]
```diff
@@ -4,28 +4,53 @@
...
+      // Maker suer more-specific versions come before Focus.
```
Comment (relud) [inline ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java]:
```suggestion
      // Maker sure more-specific versions come before Focus.
```

File: ingestion-beam/src/test/java/com/mozilla/telemetry/util/NormalizeTest.java [hunk 6]
```diff
@@ -44,4 +44,64 @@ public void channel() {
...
+    assertEquals("Mac", Normalize.os("Darwin"));
+    assertEquals("Linux", Normalize.os("Linux"));
+    assertEquals("Linux", Normalize.os("GNU/Linux"));
+    assertEquals("Linux", Normalize.os("SunOS"));
+    assertEquals("Linux", Normalize.os("FreeBSD"));
+    assertEquals("Linux", Normalize.os("GNU/kFreeBSD"));
+    assertEquals("Other", Normalize.os("AIX"));
+
+    // Mobile OS.
+
```
Comment (relud) [inline ingestion-beam/src/test/java/com/mozilla/telemetry/util/NormalizeTest.java]: this seems like it needs either at least one test or removal

Comment (jklukas) [reply ingestion-beam/src/test/java/com/mozilla/telemetry/util/NormalizeTest.java]: Totally spaced on that. Will fix.

File: ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java (hunk 4 above)
Comment (trink) [reply ingestion-beam/src/main/java/com/mozilla/telemetry/util/Normalize.java]: Yeah, they should be reordered
https://github.com/mozilla-services/lua_sandbox_extensions/pull/431

**Comment by jklukas** [comment]
> I'd recommend matching tests of the normalization logic with any in the lua implementation

I'm going to go ahead and merge with the tests we have available now, but if we find additional existing test cases that we should replicate here or we want to pursue a shared test suite, let's file a separate issue.

**Pull request merged**